sling/nlp/embedding/fact-plausibility.cc

```diff
@@ -0,0 +1,547 @@
+    }
+
+    delete optimizer_;
+  }
+
+  // Worker thread for training embedding model.
+  void Worker(int index, Network *model) override {
+    // Initialize random number generator.
+    Random rnd;
+    rnd.seed(seed_ + index);
+
+    // Premises and hypotheses for one batch.
+    std::vector<Features> premises(batch_size_);
+    std::vector<Features> hypotheses(batch_size_);
+
+    // Set up plausibility scorer.
+    Instance scorer(flow_.scorer);
+    Instance gscorer(flow_.gscorer);
+    int *premise = scorer.Get<int>(flow_.premise);
+    int *hypothesis = scorer.Get<int>(flow_.hypothesis);
+    float *logits = scorer.Get<float>(flow_.logits);
+    float *dlogits = gscorer.Get<float>(flow_.d_logits);
+    std::vector<Instance *> gradients{&gscorer};
+
+    for (;;) {
+      // Compute gradients for epoch.
+      gscorer.Clear();
+      gscorer.Set(flow_.primal, &scorer);
+      float epoch_loss = 0.0;
+      Benchmark batch_benchmark;
+
+      for (int b = 0; b < batches_per_update_; ++b) {
+        // Random sample instances for batch.
+        for (int i = 0; i < batch_size_; ++i) {
+          int sample = rnd.UniformInt(training_instances_.size());
+          Frame instance(&store_, training_instances_[sample]);
+
+          // Split instance into premise and hypothesis.
+          int num_groups = instance.Get(p_groups_).AsArray().length();
+          int heldout = rnd.UniformInt(num_groups);
+          Split(instance, &premises[i], &hypotheses[i], heldout);
+        }
+
+        // Do forward and backward propagation for each premise/hypothesis pair.
+        // Each sampled item is a positive example. Negative examples are
+        // generated by using the premise from one item and the hypothsis from
+        // another item.
+        for (int i = 0; i < batch_size_; ++i) {
+          for (int j = 0; j < batch_size_; ++j) {
+            // Check that main hypothesis is not in premise for negative
+            // examples.
+            if (i != j && Has(premises[i], hypotheses[j][0])) {
+              num_contradictions_->Increment();
+              continue;
+            }
+
+            // Set premise and hypothesis for example.
+            Copy(premises[i], premise);
+            Copy(hypotheses[j], hypothesis);
+
+            // Compute plausibility scores.
+            scorer.Compute();
+
+            // Compute accuracy.
+            batch_benchmark.add(i == j, logits[1] > logits[0]);
+
+            // Compute loss.
+            int label = i == j ? 1 : 0;
+            float loss = loss_.Compute(logits, label, dlogits);
+            epoch_loss += loss;
+
+            // Backpropagate.
+            gscorer.Compute();
+            num_training_examples_->Increment();
+          }
+        }
+      }
+
+      // Update parameters.
+      optimizer_mu_.Lock();
+      optimizer_->Apply(gradients);
+      loss_sum_ += epoch_loss;
+      loss_count_ += batches_per_update_ * batch_size_ * batch_size_;
```

Comment:
Shouldn't we omit counting those cases where the hypothesis of a -ve example was in the premise?

Reply:
Yes, otherwise the avg loss will be slightly off. I have fixed this.